src/cgroups/common.rs

```diff
@@ -167,3 +169,41 @@ pub fn create_cgroup_manager<P: Into<PathBuf>>(
         _ => bail!("could not find cgroup filesystem"),
     }
 }
+
+pub fn get_all_pids(path: &Path) -> Result<Vec<Pid>> {
+    log::debug!("scan pids in folder: {:?}", path);
+    let mut result = vec![];
+    let proc_str = OsStr::new(CGROUP_PROCS);
+    walk_path(&path, &mut |p| {
+        let file_name = p.file_name();
+        if let Some(file_name) = file_name {
+            if file_name == proc_str {
+                let file = File::open(p)?;
+                for line in BufReader::new(file).lines() {
+                    if let Ok(line) = line {
+                        result.push(Pid::from_raw(line.parse::<i32>()?))
+                    }
+                }
+            }
+        }
+        Ok(())
+    })?;
+    Ok(result)
+}
+
+fn walk_path<F>(path: &Path, c: &mut F) -> Result<()>
+where
+    F: FnMut(&Path) -> Result<()>,
+{
+    for entry in fs::read_dir(path)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if path.is_dir() {
+            walk_path(&path, c)?;
+        } else {
+            c(&path)?;
+        }
+    }
+    Ok(())
+}
```

Comment:
Couldn't we just check if the cgroup.procs file exists instead of checking the name of every file?

```
pub fn get_all_pids(path: &Path) -> Result<Vec<Pid>> {
    log::debug!("scan pids in folder: {:?}", path);
    let mut result = vec![];
    walk_path(&path, &mut result)?;

    Ok(result)
}

fn walk_path(path: &Path, result: &mut Vec<Pid>) -> Result<()> {
    let procs_path = path.join(CGROUP_PROCS);
    if procs_path.exists() {
        let file = File::open(procs_path)?;
        for line in BufReader::new(file).lines() {
            if let Ok(line) = line {
                result.push(Pid::from_raw(line.parse::<i32>()?))
            }
        }
    }

    for entry in fs::read_dir(path)? {
        let entry = entry?;
        let entry_path = entry.path();

        if entry_path.is_dir() {
            walk_path(&entry_path, result)?;
        }
    }
    Ok(())
}
```

Reply:
Are there any advantages? 
It is always necessary to traverse all files, and the file name is compared once in the process, I think it should not have any effect.

Reply:
It would reduce the number of string comparisons, but in the end it is a small thing. I am fine with it.

Reply:
thanks, I will optimize it.